### PR TITLE
Use go-events package

### PR DESCRIPTION
**Pull request by halcyonCorsair**
This addresses #1553.

TBD: Queue not converted yet

Signed-off-by: Elliot Pahl <elliot.pahl@gmail.com>

### CODE REVIEW

**Comment by halcyonCorsair** [comment]
See also #2551 and docker/go-events#24.

File: notifications/http_test.go [hunk 1]
```diff
@@ -51,7 +52,7 @@ func TestHTTPSink(t *testing.T) {
 		}
 
 		// Let caller choose the status
-		status, err := strconv.Atoi(r.FormValue("status"))
+		status, err := strconv.Atoi(r.URL.Query().Get("status"))
```
Comment (stevvooe) [inline notifications/http_test.go]: Why is this change made? `r.FormValue` is equivalent, no?

File: notifications/endpoint.go [hunk 2]
```diff
@@ -61,7 +63,7 @@ func NewEndpoint(name, url string, config EndpointConfig) *Endpoint {
 	endpoint.Sink = newHTTPSink(
 		endpoint.url, endpoint.Timeout, endpoint.Headers,
 		endpoint.Transport, endpoint.metrics.httpStatusListener())
-	endpoint.Sink = newRetryingSink(endpoint.Sink, endpoint.Threshold, endpoint.Backoff)
+	endpoint.Sink = events.NewRetryingSink(endpoint.Sink, events.NewBreaker(endpoint.Threshold, endpoint.Backoff))
```
Comment (stevvooe) [inline notifications/endpoint.go]: Looking good here.

**Comment by stevvooe** [comment]
cc @dmcgowan

File: notifications/http_test.go (hunk 1 above)
Comment (halcyonCorsair) [reply notifications/http_test.go]: This was failing, looks like due to a mistake I had made during the conversion, so I've set it back now.

**Comment by halcyonCorsair** [comment]
Updated to resolve merge conflicts.

File: notifications/metrics.go [hunk 3]
```diff
@@ -56,24 +58,36 @@ type endpointMetricsHTTPStatusListener struct {
...
 var _ httpStatusListener = &endpointMetricsHTTPStatusListener{}
 
-func (emsl *endpointMetricsHTTPStatusListener) success(status int, events ...Event) {
+func (emsl *endpointMetricsHTTPStatusListener) success(status int, event events.Event) {
 	emsl.safeMetrics.Lock()
 	defer emsl.safeMetrics.Unlock()
-	emsl.Statuses[fmt.Sprintf("%d %s", status, http.StatusText(status))] += len(events)
-	emsl.Successes += len(events)
+	evt, ok := event.(Event)
+	if !ok || (ok && evt.ID == "") {
```
Comment (stevvooe) [inline notifications/metrics.go]: Does this happen? If we are seeing this, there might be a bug.

Comment (halcyonCorsair) [reply notifications/metrics.go]: I believe this shouldn't happen.  From memory I think I added this after I accidentally sent the wrong thing through here partway through converting things.  I'll remove it.

**Comment by stevvooe** [comment]
LGTM

I think the tests are failing with a concurrency issue when setting the url on the sink for each test case: https://github.com/docker/distribution/pull/2550/files#diff-e8f917a7e7a5472ef1ca0e58b0f7775aR166.

**Pull request merged**